### ISSUE TITLE
ConcurrentModificationException in javassist.CtClassType.instrument

### ISSUE THREAD

**Issue by andrei-ivanov** [powermock/powermock#924]
Using version 1.7.4, everything works fine.
Upgrading Hibernate from 5.3.2 to 5.3.3 triggered a transitive upgrade of javassist from  
3.22.0-GA to 3.23.1-GA (I see powermock has a dependency on an even older 3.21-GA)

Using 3.23.1-GA triggers this error:
```
at org.powermock.core.classloader.MockClassLoader.loadMockClass(MockClassLoader.java:296)
	at org.powermock.core.classloader.MockClassLoader.loadModifiedClass(MockClassLoader.java:204)
	at org.powermock.core.classloader.DeferSupportingClassLoader.loadClass1(DeferSupportingClassLoader.java:89)
	at org.powermock.core.classloader.DeferSupportingClassLoader.loadClass(DeferSupportingClassLoader.java:79)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at sun.reflect.generics.factory.CoreReflectionFactory.makeNamedType(CoreReflectionFactory.java:114)
	at sun.reflect.generics.visitor.Reifier.visitClassTypeSignature(Reifier.java:125)
	at sun.reflect.generics.tree.ClassTypeSignature.accept(ClassTypeSignature.java:49)
	at sun.reflect.annotation.AnnotationParser.parseSig(AnnotationParser.java:439)
	at sun.reflect.annotation.AnnotationParser.parseClassValue(AnnotationParser.java:420)
	at sun.reflect.annotation.AnnotationParser.parseClassArray(AnnotationParser.java:724)
	at sun.reflect.annotation.AnnotationParser.parseArray(AnnotationParser.java:531)
	at sun.reflect.annotation.AnnotationParser.parseMemberValue(AnnotationParser.java:355)
	at sun.reflect.annotation.AnnotationParser.parseAnnotation2(AnnotationParser.java:286)
	at sun.reflect.annotation.AnnotationParser.parseAnnotations2(AnnotationParser.java:120)
	at sun.reflect.annotation.AnnotationParser.parseAnnotations(AnnotationParser.java:72)
	at java.lang.Class.createAnnotationData(Class.java:3521)
	at java.lang.Class.annotationData(Class.java:3510)
	at java.lang.Class.getAnnotation(Class.java:3415)
	at org.junit.internal.MethodSorter.getDeclaredMethods(MethodSorter.java:52)
	at org.junit.internal.runners.TestClass.getAnnotatedMethods(TestClass.java:45)
	at org.junit.internal.runners.MethodValidator.validateTestMethods(MethodValidator.java:71)
	at org.junit.internal.runners.MethodValidator.validateStaticMethods(MethodValidator.java:44)
	at org.junit.internal.runners.MethodValidator.validateMethodsForDefaultRunner(MethodValidator.java:50)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.validate(PowerMockJUnit44RunnerDelegateImpl.java:124)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.<init>(PowerMockJUnit44RunnerDelegateImpl.java:86)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl.<init>(PowerMockJUnit47RunnerDelegateImpl.java:42)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit49RunnerDelegateImpl.<init>(PowerMockJUnit49RunnerDelegateImpl.java:25)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.createDelegatorFromClassloader(JUnit4TestSuiteChunkerImpl.java:172)
	at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.createDelegatorFromClassloader(JUnit4TestSuiteChunkerImpl.java:48)
	at org.powermock.tests.utils.impl.AbstractTestSuiteChunkerImpl.createTestDelegators(AbstractTestSuiteChunkerImpl.java:108)
	at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.<init>(JUnit4TestSuiteChunkerImpl.java:71)
	at org.powermock.modules.junit4.common.internal.impl.AbstractCommonPowerMockRunner.<init>(AbstractCommonPowerMockRunner.java:36)
	at org.powermock.modules.junit4.PowerMockRunner.<init>(PowerMockRunner.java:34)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.junit.internal.builders.AnnotatedBuilder.buildRunner(AnnotatedBuilder.java:104)
	at org.junit.internal.builders.AnnotatedBuilder.runnerForClass(AnnotatedBuilder.java:86)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.requests.ClassRequest.getRunner(ClassRequest.java:33)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:49)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:909)
	at java.util.ArrayList$Itr.next(ArrayList.java:859)
	at javassist.CtClassType.instrument(CtClassType.java:1540)
	at org.powermock.core.transformers.impl.ClassMockTransformer.transformMockClass(ClassMockTransformer.java:65)
	at org.powermock.core.transformers.impl.AbstractMainMockTransformer.transform(AbstractMainMockTransformer.java:62)
	at org.powermock.core.classloader.MockClassLoader.loadMockClass(MockClassLoader.java:277)
	... 53 more
```
Not sure though if this is a powermock issue or a Javassist one.

**Comment by andrei-ivanov** [comment]
Looking at commit https://github.com/jboss-javassist/javassist/commit/62851d0af30cdb0c54046627585451a468aacd2a, it seems the iteration was changed to use an iterator instead of `get` for each index, so the problem was probably hidden before.

**Comment by kjmcclain** [comment]
Yes, I get the same error with the following minimal test case:

```
package org.example;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.powermock.modules.junit4.PowerMockRunner;

@RunWith(PowerMockRunner.class)
public class PowerMockRunnerTest extends Base {

	@Test
	public void test() {
	}

}

class Base {

}
```

which passes when I downgrade Hibernate from 5.3.3 to 5.3.2.

**Comment by danielfariati** [comment]
The error persists on the version 2.0.0-beta.11.

**Comment by WarrenCooper-sag** [comment]
Hi,
  Any idea when this will be fixed?  3.22.0 pre-dates this issue, but it is not portable to Open JDK (Zulu) 10. i.e. jdeps scanner reports this problematic dependency: javassist.util.proxy -> sun.misc  JDK internal API (jdk.unsupported)

  There doesn't seem to be a work-around for avoiding this exception.  I hit this issue with powermock v1.6.2 and also after upgrading v1.7.4

  I've added an issue to the Javassist repo (See referenced link below.)
-thanks

**Comment by chibash** [comment]
I've fixed it.  Please try https://github.com/jboss-javassist/javassist/releases/tag/rel_3_24_0_rc

**Comment by WarrenCooper-sag** [comment]
Thank you chibash,   3.24.0_rc passes all of our tests.

**Comment by andrei-ivanov** [comment]
Seems to work fine, tests and runtime 🙂
Thank you

**Comment by Aitozi** [comment]
When will the 3.24.0 release ? @chibash

**Comment by dlogothetis** [comment]
Any update on releasing 3.24.0?

**Comment by azee** [comment]
Looks like it has been released but not deployed to Central yet
https://github.com/jboss-javassist/javassist/releases/tag/rel_3_24_0_ga
https://mvnrepository.com/artifact/org.javassist/javassist

**Comment by chibash** [comment]
3.24.0-GA will be available from Central soon (in a few days).

**Comment by thekingn0thing** [comment]
> Looking at commit jboss-javassist/javassist@62851d0, it seems the iteration was changed to use an iterator instead of get for each index, so the problem was probably hidden before.

PowerMock creates a new constructor when modifies constructor call to be able to stab constructors calls and it leads to changing count of methods in the class. In the end, we have `ConcurrentModificationException` exception. So, as for me, it's not a hidden problem. I glad to see that the issue is fixed in 3.24.0-GA. 

Here is pull request #952 to upgrade PowerMock 2.x dependencies.

**Comment by andrei-ivanov** [comment]
That's good to hear 🙂
Now there's the question of when we'll be able to use PowerMock with JUnit 5 and Mockito 😒
